### DISCUSSION TITLE
Extend with JavaScript - How to use npm packages

### DISCUSSION THREAD

**Discussion by rifaia** [pocketbase/pocketbase#3133] [Q&A]
I tried to use the [firebase-admin](https://www.npmjs.com/package/firebase-admin) sdk to verify token from *.pb.js file(s) inside the pb_hooks directory.  I am struggling to use modules inside the *.pb.js file.  Other functions without using module works

I read in the document about [Loading Modules](https://pocketbase.io/docs/js-overview/#loading-modules), but it doesn't work. 

If someone already did, please add code or example

**Comment by ganigeorgiev** [comment]
You need to download the js module locally to use external JS modules (or have node and npm installed to download it in a `node_modules` dir).

And even if you do the above, keep in mind that it may not work because the PocketBase embedded JS engine is not a browser or node environment and any package that requires non-ES5 or some platform specific api (`fetch`, `fs`, etc.) may not work as expected.

**Comment by tigawanna** [comment]
Would it be easier to start a simple node project and build it to target es5 then copy the output

**Reply by ganigeorgiev** [reply]
Yes, transpiling is an option but again if the package relies on platform specific APIs there is no guarantee that it will work.

**Reply by tigawanna** [reply]
Thanks for the reply 🙏, will play around and maybe start some recipes of what works

**Comment by rifaia** [comment]
@ganigeorgiev Thanks for quick reply. Yes, i have downloaded using npm and it is present in `node_modules` dir aligned with `package.json` in same level of `pb_dirs`. I understand, it might not work if relies on APIs like window, fs, fetch, buffer or any other runtime specific API not part of the ES5 spec may not work!. But i am not able to load.

**Reply by ganigeorgiev** [reply]
If the module is a CJS than you should be able to register it with `require("firebase-admin/app")` but I cannot guarantee that it will work.
Alternatively, you can also try to specify the full path to the module, aka.
```
require(`${__hooks}/../path/to/your/node_modules/yourModule/...`)
```

**Reply by rifaia** [reply]
Thanks @ganigeorgiev , I will continue with [Extend with Go](https://pocketbase.io/docs/go-overview/) for now and try js while working with CJS only modules.
